source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+    {
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        return;
+
+      // Try to safely initialize the global environment
+      TapelessDrivers<ad_type, scalar_type>::initialize_global_environment(
+        n_independent_variables);
+
+      if (ensure_persistent_setting == true)
+        if (is_adolc_tapeless_number<ad_type>::value)
+          {
+            // In order to ensure that the settings remain for the entire
+            // duration of the simulation, we create a global live variable
+            // that doesn't go out of scope.
+            static ad_type num = 0.0;
+          }
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::activate_recorded_tape(
+      const types::tape_index &tape_index)
+    {
+      activate_tape(tape_index, true /*read_mode*/);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::activate_tape(
+      const types::tape_index &tape_index,
+      const bool               read_mode)
+    {
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(tape_index != types::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+          Assert(tape_index < types::max_tape_index,
+                 ExcMessage("Tape index exceeds maximum allowable value"));
+          Assert(is_registered_tape(tape_index),
+                 ExcMessage("Tape number not registered"));
+          active_tape_index = tape_index;
+          reset_touched_independent_variables();
+
+          // A tape may have been defined by a different ADHelper, so in this
+          // case we ignore the fact that any dependent variables within the
+          // current data structure have not been marked as dependents
+          if (read_mode == true)
+            {
+              reset_touched_dependent_variables(true);
+              Assert(n_touched_dependent_variables() == n_dependent_variables(),
+                     ExcMessage("Not all dependent variables have been set!"));
+            }
+        }
+      else
+        {
+          Assert(ADNumberTraits<ad_type>::is_tapeless == true,
+                 ExcInternalError());
+          active_tape_index = tape_index; // This is, in effect, a dummy value
+        }
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::set_tape_buffer_sizes(
+      const types::tape_buffer_sizes in_obufsize,
+      const types::tape_buffer_sizes in_lbufsize,
+      const types::tape_buffer_sizes in_vbufsize,
+      const types::tape_buffer_sizes in_tbufsize)
```

Comment:
this `types` namespace is inside namespace `AD`, right? I ask, because in many other contexts here, you explicitly qualify with `AD::`.